DCOS-48339: volumes form section

natmegs opened the pull request:
Add the volumes section to the jobs form.

## Testing

- Open jobs form, go to Volumes tab
- Add volumes
- Check JSON editor to make sure they appear as expected
- Try to submit with invalid volumes (missing fields, wrong field type, etc)

## Trade-offs

N/A

## Dependencies

DCOS-48339 is not complete without EE changes.

## Screenshots

<img width="1353" alt="Screen Shot 2019-05-01 at 5 01 21 PM" src="https://user-images.githubusercontent.com/19582796/57050714-ca514880-6c32-11e9-952a-2daca8000603.png">

GeorgiSTodorov commented:
- When I change the mode in the JSON, the form still shows "Read and Write" mode.
![Screenshot from 2019-05-02 12-47-34](https://user-images.githubusercontent.com/40791275/57067745-857fed00-6cd8-11e9-8ed4-98df4933c3f7.png)
![Peek 2019-05-02 12-50](https://user-images.githubusercontent.com/40791275/57067892-e3143980-6cd8-11e9-92b5-ad00416a0938.gif)

- Errors persist.

- Like in the Placement PR, if I introduce a JSON error and later fix it, I can no longer submit the form.
![Peek 2019-05-02 13-04](https://user-images.githubusercontent.com/40791275/57068589-dabcfe00-6cda-11e9-9c0e-507955c44cdf.gif)

- Error for mode doesn't disappear.
![Peek 2019-05-02 13-02](https://user-images.githubusercontent.com/40791275/57068503-929ddb80-6cda-11e9-881a-62ff296c4df1.gif)

natmegs commented:
@GeorgiSTodorov thanks for the prompt review!! I changed the mode input so that it will show `Select ...` if one of the RO/RW options are not provided. Any errors that are persisting are actually from the server - the reason for that is we cannot revalidate a rule that our client side validation didn't catch. To mitigate the issue, I've added more validation to check that `mode` is one of RO/RW. 

The invalid JSON issue actually affects the whole form... we do not receive an error change event from the JSON editor in this particular situation. It is definitely something that should be addressed but I would say it is out of scope of this particular PR.

brandonc commented on plugins/jobs/src/js/components/form/VolumesFormSection.tsx:
```diff
@@ -0,0 +1,220 @@
+        mode = "";
+      }
+
+      if (index === 0) {
```
This could be > 0 if the first volumes were SecretVolumes, right?

brandonc commented on plugins/jobs/src/js/components/form/VolumesFormSection.tsx:
```diff
@@ -0,0 +1,220 @@
+    } = this.props;
+
+    return volumes.map((volume: JobVolume | SecretVolume, index: number) => {
+      if (isFBS(volume)) {
```
Does this indicate SecretVolume? I think it must since everything gets cast as JobVolumn below. I wonder if it would be better to filter out all the SecretVolumns before the map and then avoid all the casting later? Your index === 0 thing would also not be a bug (if I understand this correctly)

```
volumes.filter((volume: JobVolume | SecretVolume) => isFBS(volume)
       .map((volume: JobVolume) => { ... }
```

natmegs replied:
If I filter the list the error indices may not match up properly in the UI - but I changed the return type on `isFBS` so that the type casting is not necessary anymore!

brandonc commented on plugins/jobs/src/js/components/form/VolumesFormSection.tsx:
```diff
@@ -0,0 +1,220 @@
+      const modeErrors = getFieldError(`job.run.volumes.${index}.mode`, errors);
+
+      return (
+        <FormRow key={index}>
```
index as key is probably insufficient since it's not permanent and unique. Is there another ID you can use?

TattdCodeMonkey commented:
@natmegs When I try to type into `Container Path` it loses focus after every key press.

natmegs commented:
@TattdCodeMonkey oof good catch! Made a bad choice for a `key` on the form row - it should be fixed now.

natmegs commented:
@Poltergeist I appreciate the thoughtful review! For 1) and 2), I think they are definitely worth addressing but I think these are out of scope for this PR. These all involve shared components that are used throughout the jobs and services forms (and likely other places too). 
For 3), my concern here is it would make validation difficult. For example, if someone adds a volume row but doesn't fill it out, this should be ignored. But if there is a default mode value, how do we know if it's truly an empty row or if the user chose it intentionally and we should be giving them a validation message to fill in host & container paths?
